File: src/lib/bridge/highlight.ts
import { invoke } from "@tauri-apps/api/core";
import type { Highlight, HighlightColor, Locator } from "$lib/domain/highlight";

export function createHighlight(args: {
  paperId: string;
  locator: Locator;
  excerpt: string;
  // Optional (RFC 0061): omit/null to create a note-/ask-only passage (no color mark).
  color?: HighlightColor | null;
  label?: string | null;
}): Promise<Highlight> {
  return invoke<Highlight>("create_highlight", {
    paperId: args.paperId, locator: args.locator, excerpt: args.excerpt,
    color: args.color ?? null, label: args.label ?? null,
  });
}

export function recolorHighlight(id: string, color: HighlightColor): Promise<void> {
  return invoke("recolor_highlight", { id, color });
}

export function setHighlightLabel(id: string, label: string | null): Promise<void> {
  return invoke("set_highlight_label", { id, label });
}

/// Set or clear a passage's note (RFC 0061).
export function setHighlightNote(id: string, note: string | null): Promise<void> {
  return invoke("set_highlight_note", { id, note });
}

export function removeHighlight(id: string): Promise<void> {
  return invoke("remove_highlight", { id });
}

/// Delete a passage outright — the mark, its note, and its conversation
/// (RFC 0079 R1.3). Resolves to the number of threads that went with it, so
/// the caller can say what it cost. `removeHighlight` above deletes only the
/// mark and leaves an unreachable thread behind; prefer this one.
export function removeAnnotation(id: string): Promise<number> {
  return invoke<number>("remove_annotation", { id });
}

export function listHighlights(paperId: string): Promise<Highlight[]> {
  return invoke<Highlight[]>("list_highlights", { paperId });
}

/// Create a highlight authored by the agent (RFC 0059 Phase 2): same shape as
/// `createHighlight`, plus the model that proposed it.
export function createAgentHighlight(args: {
  paperId: string; locator: Locator; excerpt: string; color: HighlightColor; label?: string | null; model: string;
}): Promise<Highlight> {
  return invoke<Highlight>("create_agent_highlight", {
    paperId: args.paperId, locator: args.locator, excerpt: args.excerpt,
    color: args.color, label: args.label ?? null, model: args.model,
  });
}

export function listAgentHighlights(paperId: string): Promise<Highlight[]> {
  return invoke<Highlight[]>("list_agent_highlights", { paperId });
}

File: src-tauri/src/commands/highlight.rs
//! Tauri commands: the UI front door to `HighlightService` (RFC 0058).

use crate::domain::highlight::{Highlight, HighlightAuthor, HighlightColor, Locator};
use crate::services::highlight::HighlightService;

#[tauri::command]
pub async fn create_highlight(
    service: tauri::State<'_, HighlightService>,
    paper_id: String,
    locator: Locator,
    excerpt: String,
    // Optional (RFC 0061): a null color creates a note-/ask-only passage.
    color: Option<HighlightColor>,
    label: Option<String>,
) -> Result<Highlight, String> {
    service
        .create_highlight(
            &paper_id,
            locator,
            &excerpt,
            color,
            label,
            HighlightAuthor::User,
        )
        .await
}

#[tauri::command]
pub async fn set_highlight_note(
    service: tauri::State<'_, HighlightService>,
    id: String,
    note: Option<String>,
) -> Result<(), String> {
    service.set_note(&id, note).await
}

#[tauri::command]
pub async fn recolor_highlight(
    service: tauri::State<'_, HighlightService>,
    id: String,
    color: HighlightColor,
) -> Result<(), String> {
    service.recolor(&id, color).await
}

#[tauri::command]
pub async fn set_highlight_label(
    service: tauri::State<'_, HighlightService>,
    id: String,
    label: Option<String>,
) -> Result<(), String> {
    service.set_label(&id, label).await
}

#[tauri::command]
pub async fn remove_highlight(
    service: tauri::State<'_, HighlightService>,
    id: String,
) -> Result<(), String> {
    service.remove(&id).await
}

/// Delete a passage outright: the mark, its note, and its conversation
/// (RFC 0079 R1.3). Returns how many threads went with it, which is what the
/// UI confirms against before calling.
#[tauri::command]
pub async fn remove_annotation(
    service: tauri::State<'_, HighlightService>,
    id: String,
) -> Result<usize, String> {
    service.remove_annotation(&id).await
}

#[tauri::command]
pub async fn list_highlights(
    service: tauri::State<'_, HighlightService>,
    paper_id: String,
) -> Result<Vec<Highlight>, String> {
    service.list(&paper_id).await
}

/// Agent-authored counterpart to `create_highlight` (RFC 0059). Thread
/// linkage is deferred; the agent tool passes no thread id yet.
#[tauri::command]
pub async fn create_agent_highlight(
    service: tauri::State<'_, HighlightService>,
    paper_id: String,
    locator: Locator,
    excerpt: String,
    color: HighlightColor,
    label: Option<String>,
    model: String,
) -> Result<Highlight, String> {
    service
        .create_highlight(
            &paper_id,
            locator,
            &excerpt,
            Some(color), // agent marks always carry a color
            label,
            HighlightAuthor::Agent { model },
        )
        .await
}

#[tauri::command]
pub async fn list_agent_highlights(
    service: tauri::State<'_, HighlightService>,
    paper_id: String,
) -> Result<Vec<Highlight>, String> {
    service.list_agent(&paper_id).await
}
